addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1464,3 +1539,157 @@ func (r *ClusterBootstrapReconciler) reconcileClusterProxyAndNetworkSettings(clu
+	err = r.Client.Get(r.context, key, clusterKubeConfigSecret)
+	if err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "unable to fetch cluster kubeconfig secret")
+		return err
+	}
+	if !apierrors.IsNotFound(err) {
+		err = r.removeFinalizer(clusterKubeConfigSecret)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove finalizer from cluster
+	log.Info("removing finalizer from cluster if present")
+	err = r.removeFinalizer(cluster)
+	if err != nil {
+		return err
+	}
+
+	log.Info("all finalizers have been removed from cluster resources")
+	return nil
+}
+
+func (r *ClusterBootstrapReconciler) reconcileDelete(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (ctrl.Result, error) {
+	okToRemoveFinalizers := false
+	var err error
+
+	timeOutReached := time.Now().After(cluster.GetDeletionTimestamp().Add(r.Config.ClusterDeleteTimeout))
+	if timeOutReached {
+		okToRemoveFinalizers = true
+	} else {
+		okToRemoveFinalizers, err = r.clusterIsReadyForDeletion(cluster, log)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if !okToRemoveFinalizers {
+		return ctrl.Result{RequeueAfter: constants.RequeueAfterDuration}, nil
+	}
+
+	if err = r.removeFinalizersFromClusterResources(cluster, log); err != nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ClusterBootstrapReconciler) removeFinalizer(o client.Object) error {
+	controllerutil.RemoveFinalizer(o, addontypes.AddonFinalizer)
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return r.Client.Update(r.context, o)
+	})
+}
+
+func hasPackageInstalls(ctx context.Context, remoteClient client.Client,
+	cluster *clusterapiv1beta1.Cluster, namespace string, packages []*runtanzuv1alpha3.ClusterBootstrapPackage,
+	log logr.Logger) bool {
+
+	for _, pkg := range packages {
+		pkgInstallName := util.GeneratePackageInstallName(cluster.Name, pkg.RefName)
+		if packageInstallExists(ctx, pkgInstallName, namespace, remoteClient, log) {
+			return true
+		}
+	}
+	return false
+}
+
+func packageInstallExists(ctx context.Context, pkgInstallName, pkgInstallNamespace string,
+	remoteClient client.Client, log logr.Logger) bool {
+
+	pkgInstall := &kapppkgiv1alpha1.PackageInstall{}
+	if err := remoteClient.Get(ctx, client.ObjectKey{Name: pkgInstallName, Namespace: pkgInstallNamespace}, pkgInstall); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false
+		}
+		log.Error(err, "could not verify if package install is present")
+		return true
+	}
+	return true
+}
+
+func (r *ClusterBootstrapReconciler) removeAdditionalPackageInstalls(remoteClient client.Client, cluster *clusterapiv1beta1.Cluster, clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap, log logr.Logger) error {
+	// Removes all additional package install CRs from the cluster
+
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil {
+		return err
+	}
+	if clusterBootstrap == nil {
```

Comment:
Do we need this check? If `clusterBootstrap == nil`, I guess it will return directly in line 1674.